### PR TITLE
Arithmetic in the WHERE clause

### PR DESCRIPTION
**Pull request by andytill**
RFC: https://github.com/basho/rfc/pull/51
riak_ql: https://github.com/basho/riak_ql/pull/165
riak_kv: https://github.com/basho/riak_kv/pull/1611

```sql
SELECT * FROM mytable
WHERE myts >= 10000 AND myts <= 10000 + 10m
```

### CODE REVIEW

File: src/riak_ql_ddl.erl [hunk 1]
```diff
@@ -378,17 +381,22 @@ is_query_valid_result(Res) ->
       true, Res).
 
 -spec are_filters_valid(module(), [filter()]) -> true | {false, [query_syntax_error()]}.
-are_filters_valid(Mod, Where) ->
-    Errors = fold_where_tree(fun(_, Clause, Acc) ->
-                                     is_filters_field_valid(Mod, Clause, Acc)
-                             end, [], Where),
+    are_filters_valid(Mod, Where) ->
```
Comment (paegun) [inline src/riak_ql_ddl.erl]: lint: remove leading spaces on function head (pin to left margin (^)).

Comment (andytill) [reply src/riak_ql_ddl.erl]: Fixed in 42ac4b1

File: test/parser_select_tests.erl [hunk 2]
```diff
@@ -567,3 +567,12 @@ single_line_comment_in_multiline_ctrl_select_test() ->
...
 
+select_with_arithmetic_in_where_clause_test() ->
+    Query_sql =
+        "SELECT * FROM mytab "
+        "WHERE a = 10 + 1",
+    {select, Parsed_query} = riak_ql_parser:ql_parse(riak_ql_lexer:get_tokens(Query_sql)),
+    ?assertEqual(
+        {where, [{'=', <<"a">>, {'+',{'integer', 10},{'integer',1}}}]},
+        proplists:lookup(where, Parsed_query)
+    ).
```
Comment (paegun) [inline test/parser_select_tests.erl:602]: When I reached this point, I had doubts that the predicate flipped about the equals sign would work (`... WHERE 10 + 1 = a`) but it does. This case is covered in riak_kv tests, but the feature backing it is actually here in riak_ql, so would IMHO be more appropriate to have such a test here too.

File: src/riak_ql_ddl.erl [hunk 3]
```diff
@@ -379,16 +382,21 @@ is_query_valid_result(Res) ->
...
+        end, [], Where),
     case Errors of
         [] -> true;
         _  -> {false, Errors}
     end.
 
-%% the terminal case of "a = 2"
-is_filters_field_valid(Mod, {Op, Field, {RHS_type, RHS_Val}}, Acc1) ->
+is_filters_field_valid(_, {Op, {identifier,FieldName}, _}, Acc1) when (Op == '/' orelse Op == '+' orelse Op == '-' orelse Op == '*') ->
+    [{arithmetic_on_identifier, FieldName} | Acc1];
```
Comment (paegun) [inline src/riak_ql_ddl.erl:397]: It seems that there should be a call to `Mod:is_field_valid/1` for the identifier when it is w/i an arithmetic operation as well.

File: test/parser_select_tests.erl (hunk 2 above)
Comment (paegun) [reply test/parser_select_tests.erl:602]: This test also clearly (w/ some brain power and reading) shows that riak_kv is performing the functionality of resolving literal-op-literal arithmetic before issuing the query to the cluster.

**Comment by paegun** [comment]
Please fix the merge conflict, a plus one is awaiting :)

**Comment by paegun** [comment]
+1

File: src/riak_ql_ddl.erl (hunk 3 above)
Comment (macintux) [reply src/riak_ql_ddl.erl:397]: Good catch:

```
riak-shell(7)>select * from a1 where time = 10000 + 10m;                        No rows returned.
riak-shell(8)>select * from a1 where foo = 10000 + 10m;
UNEXPECTED ERROR - if you have logging on please send your logfile to Basho: Error processing incoming message: error:function_clause:[{riak_ql_table_a1_224106348230695526464726407567329086520, get_field_type, [[<< foo >>]], [{file,[]}, {line,5}]}, {riak_kv_qry_compiler, resolve_expressions_folder, 2, [{file, src/riak_kv_qry_compiler.erl }, {line,1183}]}, {riak_kv_qry_compiler, '-resolve_expressions/2-fun-0-', 4, [{file, src/riak_kv_qry_compiler.erl }, {line,1169}]}, {riak_ql_ddl, mapfold_where_tree, 3, [{file, src/riak_ql_ddl.erl }, {line,553}]}, {riak_kv_qry_compiler, compile_where_clause, 2, [{file, src/riak_kv_qry_compiler.erl }, {line,168}]}, {riak_kv_qry, do_select,2, [{file, src/riak_kv_qry.erl }, {line,242}]}, {riak_kv_ts_api, query,2, [{file, src/riak_kv_ts_api.erl }, {line,230}]}, {riak_kv_ts_svc, sub_tsqueryreq,4, [{file, src/riak_kv_ts_svc.erl }, {line,446}]}]```

Comment (andytill) [reply src/riak_ql_ddl.erl:397]: Fixed in 4b7969a and corresponding riak_kv commit.

File: src/riak_ql_ddl.erl [hunk 4]
```diff
@@ -379,16 +382,21 @@ is_query_valid_result(Res) ->
...
+                {Clause, is_filters_field_valid(Mod, Clause, Acc)}
+        end, [], Where),
     case Errors of
         [] -> true;
         _  -> {false, Errors}
     end.
 
-%% the terminal case of "a = 2"
-is_filters_field_valid(Mod, {Op, Field, {RHS_type, RHS_Val}}, Acc1) ->
+is_filters_field_valid(_, {Op, {identifier,FieldName}, _}, Acc1) when (Op == '/' orelse Op == '+' orelse Op == '-' orelse Op == '*') ->
```
Comment (macintux) [inline src/riak_ql_ddl.erl:396]: Very close to a thumbs up, but I do have to object to this function name. `is_` functions should return true/false (or maybe a false tuple like `are_filters_valid`, maybe).

Comment (macintux) [reply src/riak_ql_ddl.erl:396]: And perhaps this is out of scope for this PR, but I can't help but wonder, given the multiple places where we're doing a guard against various operations, whether we could turn them into macros.

```
-define(IS_MATH_OP(Op), Op == '/' orelse Op == '+' orelse Op == '-' orelse Op == '*').
-define(IS_COMPAR_OP(Op), Op == '=' orelse Op == '!=' orelse Op == '>' orelse Op == '<' orelse Op == '>=' orelse Op == '<=').
```

**Comment by macintux** [comment]
+1

File: src/riak_ql_ddl.erl (hunk 4 above)
Comment (macintux) [reply src/riak_ql_ddl.erl:396]: Agreed with Andy that both these are out of scope for the current work.

**Comment by paegun** [comment]
+1

**Pull request merged**